### ISSUE TITLE
Grafana send multiple repeat alerting messages

### ISSUE THREAD

**Issue by SperChung-11** [grafana/grafana#35083]
**What happened**: We have set an alert rule for a dashboard on Grafana.  While the alerting rule was triggered, we found that the Grafana send two alerting messages to oncall randomly.

**What you expected to happen**: the Grafana send one alerting message to oncall.

**How to reproduce it (as minimally and precisely as possible)**: Set an alerting rule and trigger an alarm. 

**Anything else we need to know?**: There are two grafana instances which are using the same database instance (mysql 5.7.31)

**Environment**:
Grafana version: 7.0.3
- Data source type & version: Mysql 5.7.31
- OS Grafana is installed on: docker
- User OS & Browser: Euler OS 2.7
- Grafana plugins: elasticsearch 
- Others:
![alerting](https://user-images.githubusercontent.com/74529952/120444150-f0b9b300-c3b9-11eb-80d4-0019f6e186b5.PNG)

**Comment by SperChung-11** [comment]
> @SperChung-11 you mentioned that you are using two grafana instances. Are you running two instances against the same grafana database? or do you mean that you are running two grafana instances against the same database as a datasource in grafana?

---- hello，it means that two instances are running against the same grafana database

**Comment by mckn** [comment]
> > @SperChung-11 you mentioned that you are using two grafana instances. Are you running two instances against the same grafana database? or do you mean that you are running two grafana instances against the same database as a datasource in grafana?
> 
> ---- hello，it means that two instances are running against the same grafana database

Ah, okay - so that is the reason why you get multiple alerting messages because the evaluation of the rules is being done in the grafana instance.

**Comment by SperChung-11** [comment]
> > > @SperChung-11 you mentioned that you are using two grafana instances. Are you running two instances against the same grafana database? or do you mean that you are running two grafana instances against the same database as a datasource in grafana?
> > 
> > 
> > ---- hello，it means that two instances are running against the same grafana database
> 
> Ah, okay - so that is the reason why you get multiple alerting messages because the evaluation of the rules is being done in the grafana instance.

yeah...But in fact, it should execute evaluation of the rules one time, no matter how many grafana instances are running. So what should i do to avoid this problem ?

**Comment by mckn** [comment]
hmm, I'm not sure. I will see if @kylebrandt or @papagian could help out to figure out if this can be resolved.

**Comment by malejpavouk** [comment]
Running two instances of Grafana agains a shared Grafana database (and same datasources) is standard HA setup. The alerts imho should not be fired twice (we have a very similar issue with 8.0.0).

**Comment by kylebrandt** [comment]
Being worked on in https://github.com/grafana/grafana/pull/37920/files

**Comment by malejpavouk** [comment]
@kylebrandt I do not see any assignment to a future release (8.2 or 8.3). Is there any timeline, when we can expect that the HA of alerting will be finalized/released? For us, the absence of HA in alerting is (as far as I know) the last big obstacle for full adoption and migration from G7 alerting.

**Comment by gotjosh** [comment]
Hello @malejpavouk, support for HA landed in 8.2 but has been waiting on documentation for a while.

This is now documented at https://grafana.com/docs/grafana/next/administration/set-up-for-high-availability/#alerting - let us know if you have any other questions or concerns and feel free to re-open a new issue.

**Comment by malejpavouk** [comment]
@gotjosh the issue with K8s setup is that you do not know the IP address of other instances upfront (as these are assigned at pod startup). Also the number of instances and their IPs may change as the existing instances are taken down (explicitly or by orchestrator). So either you need to perform periodic service discovery or to use fixed DNS records (I guess that in that case, you must deploy Grafana in a Statefulset). 

This is the guidance that we are missing. Getting IP of the local pod is trivial, getting all >live< instances at runtime (and possibly reconfigure) is a different beast.

**Comment by gotjosh** [comment]
> So either you need to perform periodic service discovery or to use fixed DNS records (I guess that in that case, you must deploy Grafana in a Statefulset)

The clustering mechanism is able to perform basic DNS service discovery via SRV records. So there's potential for a [headless service](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services) here. Going via the  statefulset route is ideal given the DNS records for each pod are static e.g. `name-0.cluster.local, name-1.cluster.local, name-2.cluster.local` but you can also make it work with a deployment even if IPs change as we're using Gossip to propagate peers. 

I recognise that this might be the guidance you're missing, but I don't think it's the guidance we're able to provide. We have a guide on deploying Grafana in Kubernetes, but an HA setup, in Kubernetes is an entirely different topic. A wide and varied one.

How you choose to achieve what's required should be left to the reader.
